lib/rmg/TreasurePlacer.cpp

```diff
@@ -0,0 +1,795 @@
+				
+				return obj;
+			};
+			
+			possibleObjects.push_back(oi);
+			
+			oi.generateObject = [i, randomAppearance, this, generateArtInfo]() -> CGObjectInstance *
+			{
+				auto factory = VLC->objtypeh->getHandlerFor(Obj::SEER_HUT, randomAppearance);
+				auto obj = (CGSeerHut *) factory->create(ObjectTemplate());
+				obj->rewardType = CGSeerHut::RESOURCES;
+				obj->rID = Res::GOLD;
+				obj->rVal = generator.getConfig().questRewardValues[i];
+				
+				obj->quest->missionType = CQuest::MISSION_ART;
+				ArtifactID artid = *RandomGeneratorUtil::nextItem(generator.getQuestArtsRemaning(), generator.rand);
+				obj->quest->m5arts.push_back(artid);
+				obj->quest->lastDay = -1;
+				obj->quest->isCustomFirst = obj->quest->isCustomNext = obj->quest->isCustomComplete = false;
+				
+				generator.banQuestArt(artid);
+				
+				this->questArtZone->getModificator<TreasurePlacer>()->possibleObjects.push_back(generateArtInfo(artid));
+				
+				return obj;
+			};
+			
+			possibleObjects.push_back(oi);
+		}
+	}
+}
+
+bool TreasurePlacer::isGuardNeededForTreasure(int value)
+{
+	return zone.getType() != ETemplateZoneType::WATER && value > minGuardedValue;
+}
+
+std::vector<ObjectInfo*> TreasurePlacer::prepareTreasurePile(const CTreasureInfo& treasureInfo)
+{
+	std::vector<ObjectInfo*> objectInfos;
+	int maxValue = treasureInfo.max;
+	int minValue = treasureInfo.min;
+	
+	const ui32 desiredValue = generator.rand.nextInt(minValue, maxValue);
+	
+	int currentValue = 0;
+	bool hasLargeObject = false;
+	while(currentValue <= (int)desiredValue - 100) //no objects with value below 100 are available
+	{
+		auto oi = getRandomObject(desiredValue, currentValue, maxValue, !hasLargeObject);
+		if(!oi) //fail
+			break;
+		
+		if(oi->templ.isVisitableFromTop())
+		{
+			objectInfos.push_back(oi);
+		}
+		else
+		{
+			if(hasLargeObject)
+			{
+				//only one large object per treasure pile
+				continue;
+			}
+			else
+			{
+				objectInfos.insert(objectInfos.begin(), oi); //large object shall at first place
+				hasLargeObject = true;
+			}
+		}
+		
+		//remove from possible objects
+		assert(oi->maxPerZone);
+		oi->maxPerZone--;
+		
+		currentValue += oi->value;
+	}
+	
+	return objectInfos;
+}
+
+rmg::Object TreasurePlacer::constuctTreasurePile(const std::vector<ObjectInfo*> & treasureInfos)
+{
+	rmg::Object rmgObject;
+	for(auto & oi : treasureInfos)
+	{
+		auto blockedArea = rmgObject.getArea();
+		auto accessibleArea = rmgObject.getAccessibleArea();
+		if(rmgObject.instances().empty())
+			accessibleArea.add(int3());
+		
+		auto * object = oi->generateObject();
+		object->appearance = oi->templ;
+		auto & instance = rmgObject.addInstance(*object);
+
+		do
+		{
+			if(accessibleArea.empty())
+			{
+				//fail - fallback
+				rmgObject.clear();
+				return rmgObject;
+			}
+			
+			int3 nextPos = *RandomGeneratorUtil::nextItem(accessibleArea.getTiles(), generator.rand);
+			instance.setPosition(nextPos - rmgObject.getPosition());
+			
```

Comment:
Here is the tricky part. You should check the minimal distance for every new object instance placed, not only the treasure pile starting point. At least this is how I estimted the distance values.
Otherwise one treasure pile might grow in the direction of the other and close the distance.

Reply:
It's easily doable, but it can affect performance.

Reply:
Actually distance is computed for all tiles blocked by object (see ObjectManager.cpp, 53).
But yes, for optimisation purposes during finding position, only initial position of treasure pile is checked